### PR TITLE
Manage retention of partial snapshots in SLM (Simple version)

### PR DESCRIPTION
**Pull request by AthenaEryma**
Currently, partial snapshots will eventually build up unless they are
manually deleted. Partial snapshots may be useful if there is not a more
recent successful snapshot, but should eventually be deleted if they are
no longer useful.

With this change, partial snapshots are deleted using the following
strategy: PARTIAL snapshots will be kept until the configured
expire_after period has passed, if present, and then be deleted. If
there is no configured expire_after in the retention policy, then they
will be deleted if there is at least one more recent successful snapshot
from this policy (as they may otherwise be useful for troubleshooting
purposes). Partial snapshots are not counted towards either min_count or
max_count.

Implements the "simple" strategy described in https://github.com/elastic/elasticsearch/issues/46988
Relates to https://github.com/elastic/elasticsearch/issues/43663

Labelled `>non-issue` as SLM retention hasn't shipped yet, so this doesn't need to be in the release notes.

### CODE REVIEW

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/slm/SnapshotRetentionConfigurationTests.java [hunk 1]
```diff
@@ -103,13 +104,21 @@ public void testMaximum() {
     }
 
     public void testFailuresDeletedIfExpired() {
+        assertUnsuccessfulDeletedIfExpired(true);
+    }
+
+    public void testPartialsDeletedIfExpired() {
+        assertUnsuccessfulDeletedIfExpired(true);
```
Comment (dakrone) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/slm/SnapshotRetentionConfigurationTests.java]: Should this be `false`? I think this is identical to the `testFailuresDeletedIfExpired` test above?

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SLMSnapshotBlockingIntegTests.java [hunk 2]
```diff
@@ -268,7 +284,9 @@ public void testBasicFailureRetention() throws Exception {
                     assertEquals(ClusterHealthStatus.RED, client().admin().cluster().prepareHealth().get().getStatus()),
                 30, TimeUnit.SECONDS);
 
-            final String masterNode = blockMasterFromFinalizingSnapshotOnIndexFile(REPO);
+            final String masterNode;
+            masterNode = blockMasterFromFinalizingSnapshotOnIndexFile(REPO);
```
Comment (dakrone) [inline x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SLMSnapshotBlockingIntegTests.java]: What was this change for?

**Review by dakrone — APPROVED** [review]
I left a couple of comments, related to the tests. This LGTM otherwise.

I think we also need to add documentation to the `slm-retention.asciidoc` file (it can be a separate PR) about exactly how our retention treats FAILURE and PARTIAL snapshots.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/slm/SnapshotRetentionConfigurationTests.java (hunk 1 above)
Comment (gwbrown) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/slm/SnapshotRetentionConfigurationTests.java]: Nice catch, thanks!

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SLMSnapshotBlockingIntegTests.java (hunk 2 above)
Comment (gwbrown) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SLMSnapshotBlockingIntegTests.java]: I briefly experimented with only blocking data nodes to induce `PARTIAL` snapshots, but it didn't work out - this was an artifact of that, I guess I missed cleaning it up. Put it back now.

**Pull request merged**